File: leetcode_2559/CountVowelStringsInRanges.java
package leetcode_2559;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

class CountVowelStringsInRanges {
    public int[] vowelStrings(String[] words, int[][] queries) {
        Set<Character> vowels = new HashSet<>(Arrays.asList('a', 'e', 'i', 'o', 'u'));
        int n = words.length;
        int [] prefix = new int[n + 1];
        int prev = 0;

        for(int i=0; i<n; i++)
        {
            if(vowels.contains(words[i].charAt(0)) && vowels.contains(words[i].charAt(words[i].length() - 1)))
                prev++;
            prefix[i + 1] = prev;
        }

        int [] res = new int [queries.length];

        for(int i=0; i<queries.length; i++)
        {
            res[i] = prefix[queries[i][1] + 1] - prefix[queries[i][0]];
        }

        return res;
    }

    public static void main(String[] args) {
        CountVowelStringsInRanges sol = new CountVowelStringsInRanges();
        String [] words = {"aba","bcb","ece","aa","e"};
        int [][] queries = {{0, 2}, {1, 4}, {1, 1}};
        int [] res = sol.vowelStrings(words, queries);
        for(int i : res)
            System.out.print(i + " ");
    }
}

File: leetcode_2559/CountVowelStringsInRanges.py
class CountVowelStringsInRanges:
    def vowelStrings(self, words: list[str], queries: list[list[int]]) -> list[int]:
        vowels = set("aeiou")
        prefix_cnt = [0] * (len(words) + 1)
        prev = 0
        
        for i, w in enumerate(words):
            if w[0] in vowels and w[-1] in vowels:
                prev += 1
            prefix_cnt[i + 1] = prev
        
        res = [0] * len(queries)
        for i, q  in enumerate(queries):
            l, r = q
            res[i] = prefix_cnt[r + 1] - prefix_cnt[l]
        
        return res

def main():
    v = CountVowelStringsInRanges()
    words = ["aba","bcb","ece","aa","e"]
    queries = [[0, 2], [1, 4], [1, 1]]
    print(v.vowelStrings(words, queries))
    
if __name__ == "__main__":
    main()

